yetanotherco/aligned_layer
# feat: add user error metrics

# Add User Error Metrics
> [!WARNING]
> This PR requires the deploy of new metrics/dashboards


## Description
This PR adds a new panel to our batcher Grafana dashboard.
- User Error Count.

### User Error Count
Tracks the occurrence of the different user errors in the batcher. The error types included are:
- `InvalidChainId`
- `InvalidSignature`
- `ProofTooLarge`
- `DisabledVerifier`
- `RejectedProof`
- `InsufficientBalance`
- `InvalidNonce`
- `EthRpcError`
- `InvalidMaxFee`
- `AddToBatchError`
- `InvalidReplacementMessage`

<img width="592" alt="image" src="https://github.com/user-attachments/assets/e074d601-2930-4acd-8d96-20857ee8e408">


## How to test
Run:
```
make anvil_start_with_block_time
make batcher_start_local
make run_metrics
```

Go to `localhost:3000` and you should see the new panel in the batcher dashboard.

Then, you can modify the task sender in `batcher/aligned/src/main.rs` to trigger the different errors in the batcher. For example, you can modify the `chain_id` with an invalid value:
```Rust
//let chain_id = get_chain_id(eth_rpc_url.as_str()).await?;
let chain_id: u64 = 4;
```
Or change the `public_input`, to get a `rejected_proof` error:
```Rust
VerificationData {
        proving_system,
        proof,
        // pub_input,
        pub_input: None,
        verification_key,
        vm_program_code,
        proof_generator_addr,
    }
```

and run:
```
make batcher_send_risc0_burst
```

## Type of change


## Checklist

  - [Link to the open issue addressing it](https://github.com/yetanotherco/aligned_layer/issues/1369)
